x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/form/deserializer_and_serializer.test.ts

```diff
@@ -0,0 +1,183 @@
+
+const unknownValue = { some: 'value' };
+
+const populateWithUnknownEntries = (v: unknown) => {
+  if (isObject(v)) {
+    for (const key of Object.keys(v)) {
+      if (key === 'require' || key === 'include' || key === 'exclude') continue; // this will generate an invalid policy
+      populateWithUnknownEntries(v[key]);
+    }
+    v.unknown = unknownValue;
+    return;
+  }
+  if (Array.isArray(v)) {
+    v.forEach(populateWithUnknownEntries);
+  }
+};
+
+const originalPolicy: SerializedPolicy = {
+  name: 'test',
+  phases: {
+    hot: {
+      actions: {
+        rollover: {
+          max_age: '1d',
+          max_size: '10gb',
+          max_docs: 1000,
+        },
+        forcemerge: {
+          index_codec: 'best_compression',
+          max_num_segments: 22,
+        },
+        set_priority: {
+          priority: 1,
+        },
+      },
+      min_age: '12ms',
+    },
+    warm: {
+      actions: {
+        shrink: { number_of_shards: 12 },
+        allocate: {
+          number_of_replicas: 3,
+        },
+        set_priority: {
+          priority: 10,
+        },
+        migrate: { enabled: false },
+      },
+    },
+    cold: {
+      actions: {
+        allocate: {
+          number_of_replicas: 12,
+          require: { test: 'my_value' },
+          include: { test: 'my_value' },
+          exclude: { test: 'my_value' },
+        },
+        freeze: {},
+        set_priority: {
+          priority: 12,
+        },
+      },
+    },
+    delete: {
+      actions: {
+        delete: {
+          delete_searchable_snapshot: true,
+        },
+        wait_for_snapshot: {
+          policy: 'test',
+        },
+      },
+    },
+  },
+};
+
+describe('deserializer and serializer', () => {
+  let policy: SerializedPolicy;
+  let serializer: ReturnType<typeof createSerializer>;
+  let formInternal: FormInternal;
+
+  // So that we can modify produced form objects
+  beforeAll(() => setAutoFreeze(false));
+  // This is the default in dev, so change back to true (https://github.com/immerjs/immer/blob/master/docs/freezing.md)
+  afterAll(() => setAutoFreeze(true));
+
+  beforeEach(() => {
+    policy = cloneDeep(originalPolicy);
+    serializer = createSerializer(policy);
+    formInternal = deserializer(policy);
+  });
+
+  it('preserves any unknown policy settings', () => {
+    // We populate all levels of the policy with entries our UI does not know about
+    populateWithUnknownEntries(policy);
+
+    const copyOfPolicy = cloneDeep(policy);
+
+    expect(serializer(formInternal)).toEqual(policy);
+
+    // Assert that the policy we passed in is unaltered after deserialization and serialization
+    expect(policy).toEqual(copyOfPolicy);
+  });
+
+  it('removes all phases if they were disabled in the form', () => {
+    formInternal._meta.warm.enabled = false;
+    formInternal._meta.cold.enabled = false;
+    formInternal._meta.delete.enabled = false;
+
+    expect(serializer(formInternal)).toEqual({
+      ...policy,
```

Comment:
This line makes me read the definition of `policy` to understand what we're spreading and what will be retained. But it looks like the only thing we're retaining is the `name` property. I would find this easier to read if we didn't spread, and instead just hardcoded the `name`. WDYT?

```js
    expect(serializer(formInternal)).toEqual({
      name: 'test',
      phases: {
        hot: policy.phases.hot, // We expect to see only the hot phase
      },
    });
```